controllers/agentcluster_controller.go

```diff
@@ -84,17 +90,63 @@ func (r *AgentClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request
 		return ctrl.Result{}, client.IgnoreNotFound(err)
 	}
 
+	patchHelper, err := patch.NewHelper(agentCluster, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	defer func() {
+		if err = patchHelper.Patch(ctx, agentCluster); err != nil {
+			log.WithError(err).Errorf("failed patching AgentCluster")
+		}
+		log.Info("AgentCluster Reconcile ended")
+	}()
+
+	// AgentCluster is being deleted
+	if !agentCluster.DeletionTimestamp.IsZero() {
+		if err = r.handleDeletion(ctx, agentCluster); err != nil {
+			log.WithError(err).Errorf("failed to remove AgentCluster")
+			return ctrl.Result{}, err
+		}
+	}
+
+	if paused := agentCluster.Annotations[clusterv1.PausedAnnotation]; paused == "true" {
+		log.Info("Skipping reconcile of AgentCluster as it's paused, but orphan its resources")
+		if err := r.orphanResources(ctx, agentCluster); err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+
+	if !controllerutil.ContainsFinalizer(agentCluster, agentClusterFinalizer) {
+		controllerutil.AddFinalizer(agentCluster, agentClusterFinalizer)
+	}
+
 	// If the agentCluster has no reference to a ClusterDeployment, create one
-	if agentCluster.Status.ClusterDeploymentRef.Name == "" {
+	clusterDeploymentName := agentCluster.Annotations[ClusterDeploymentRefNameAnnotation]
+	clusterDeploymentNamespace := agentCluster.Annotations[ClusterDeploymentRefNamespaceAnnotation]
```

Comment:
Instead of making the status fields obsolete and then having two places we need to maintain this data I think it would be better to put a label on the `ClusterDeployment` and use that to discover our cluster if we don't have `Status.ClusterDeploymentRef`

Reply:
Great idea, thank you! Modified as suggested